qiskit/transpiler/passes/optimization/optimize_1q_commutation.py

```diff
@@ -0,0 +1,227 @@
+
+NOTE: These rules are _symmetric_, so that they may be applied in reverse.
+"""
+
+
+class Optimize1qGatesSimpleCommutation(TransformationPass):
+    """
+    Optimizes 1Q gate strings interrupted by 2Q gates by commuting the components and re-
+    synthesizing the results.  The commutation rules are stored in `commutation_table`.
+
+    NOTE: In addition to those mentioned in `commutation_table`, this pass has some limitations:
+          + Does not handle multiple commutations in a row without intermediate progress.
+          + Can only commute into positions where there are pre-existing runs.
+          + Does not exhaustively test all the different ways commuting gates can be assigned to
+            either side of a barrier to try to find low-depth configurations.  (This is particularly
+            evident if all the gates in a run commute with both the predecessor and the successor
+            barriers.)
+    """
+
+    # NOTE: A run from `dag.collect_1q_runs` is always nonempty, so we sometimes use an empty list
+    #       to signify the absence of a run.
+
+    def __init__(self, basis=None, run_to_completion=False):
+        """
+        Args:
+            basis (list[str]): See also `Optimize1qGatesDecomposition`.
+        """
+        super().__init__()
+
+        self._basis = basis
+        self._optimize1q = Optimize1qGatesDecomposition(basis)
+        self._run_to_completion = run_to_completion
+
+    @staticmethod
+    def _find_adjoining_run(dag, runs, run, front=True):
+        """
+        Finds the run which abuts `run` from the front (or the rear if `front == False`), separated
+        by a blocking node.
+
+        Returns a pair of the abutting multi-qubit gate and the run which it separates from this
+        one. The next run can be the empty list `[]` if it is absent.
+        """
+        edge_node = run[0] if front else run[-1]
+        blocker = next(dag.predecessors(edge_node) if front else dag.successors(edge_node))
+        possibilities = dag.predecessors(blocker) if front else dag.successors(blocker)
+
+        adjoining_run = []
+        for possibility in possibilities:
+            if isinstance(possibility, DAGOpNode) and possibility.qargs == edge_node.qargs:
+                adjoining_run = next((run for run in runs if possibility in run), [])
+                break
+
+        return (blocker, adjoining_run)
+
+    @staticmethod
+    def _commute_through(blocker, run, front=True):
+        """
+        Pulls `DAGOpNode`s from the front of `run` (or the back, if `front == False`) until it
+        encounters a gate which does not commute with `blocker`.
+
+        Returns a pair of lists whose concatenation is `run`.
+        """
+        run_clone = copy(run)
+
+        commuted = []
+        preindex, commutation_rule = None, None
+        if isinstance(blocker, DAGOpNode):
+            preindex = None
+            for i, q in enumerate(blocker.qargs):
+                if q == run[0].qargs[0]:
```

Comment:
I ran into a bug here when `run` is empty. Here's a minimal circuit to reproduce:
```py
qc = QuantumCircuit.from_qasm_str('''
OPENQASM 2.0;
include "qelib1.inc";
qreg q[3];
creg m[3];

cx q[0],q[1];
sx q[0];
rz(pi/2) q[1];
cx q[0],q[1];
rz(2.3082985) q[0];
sx q[1];
cx q[2],q[1];
sx q[2];
rz(-pi) q[2];
rz(-pi/2) q[1];

''')

There are other places in the code where `run[0]` is accessed too so maybe worth checking all of them
```